pallets/schemas/src/lib.rs

```diff
@@ -381,11 +445,204 @@ pub mod pallet {
+		/// * [`Error::InvalidSetting`] - Invalid setting is provided
+		///
+		#[pallet::call_index(7)]
+		#[pallet::weight(
+			match schema_name {
+				Some(_) => T::WeightInfo::create_schema_v3(model.len() as u32 + settings.len() as u32),
+				None => T::WeightInfo::create_schema_v2(model.len() as u32 + settings.len() as u32)
+			}
+		)]
+		pub fn create_schema_v3(
+			origin: OriginFor<T>,
+			model: BoundedVec<u8, T::SchemaModelMaxBytesBoundedVecLimit>,
+			model_type: ModelType,
+			payload_location: PayloadLocation,
+			settings: BoundedVec<SchemaSetting, T::MaxSchemaSettingsPerSchema>,
+			schema_name: Option<SchemaNamePayload>,
+		) -> DispatchResult {
+			let sender = ensure_signed(origin)?;
+
+			let (schema_id, schema_name) = Self::create_schema_for(
+				model,
+				model_type,
+				payload_location,
+				settings,
+				schema_name,
+			)?;
 
 			Self::deposit_event(Event::SchemaCreated { key: sender, schema_id });
+			if let Some(inner_name) = schema_name {
+				Self::deposit_event(Event::SchemaNameCreated {
+					schema_id,
+					name: inner_name.get_combined_name(),
+				});
+			}
 			Ok(())
 		}
+
+		/// Propose to create a schema name.  Creates a proposal for council approval to create a schema name
+		///
+		#[pallet::call_index(8)]
+		#[pallet::weight(T::WeightInfo::propose_to_create_schema_name())]
+		pub fn propose_to_create_schema_name(
+			origin: OriginFor<T>,
+			schema_id: SchemaId,
+			schema_name: SchemaNamePayload,
+		) -> DispatchResult {
+			let proposer = ensure_signed(origin)?;
+
+			ensure!(
+				schema_id <= Self::get_current_schema_identifier_maximum(),
+				Error::<T>::SchemaIdDoesNotExist
+			);
+
+			let proposal: Box<T::Proposal> = Box::new(
+				(Call::<T>::create_schema_name_via_governance { schema_id, schema_name }).into(),
+			);
+			T::ProposalProvider::propose_with_simple_majority(proposer, proposal)?;
+			Ok(())
+		}
+
+		/// Assigns a name to a schema without any name
+		///
+		/// # Events
+		/// * [`Event::SchemaCreated`]
+		///
+		/// # Errors
+		/// * [`Error::LessThanMinSchemaModelBytes`] - The schema's length is less than the minimum schema length
+		/// * [`Error::ExceedsMaxSchemaModelBytes`] - The schema's length is greater than the maximum schema length
+		/// * [`Error::InvalidSchema`] - Schema is malformed in some way
+		/// * [`Error::SchemaCountOverflow`] - The schema count has exceeded its bounds
+		#[pallet::call_index(9)]
+		#[pallet::weight(T::WeightInfo::create_schema_name_via_governance())]
+		pub fn create_schema_name_via_governance(
+			origin: OriginFor<T>,
+			schema_id: SchemaId,
+			schema_name: SchemaNamePayload,
+		) -> DispatchResult {
+			T::CreateSchemaViaGovernanceOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				schema_id <= Self::get_current_schema_identifier_maximum(),
+				Error::<T>::SchemaIdDoesNotExist
+			);
+			let parsed_name = SchemaName::try_parse::<T>(schema_name, true)?;
+			SchemaNameToIds::<T>::try_mutate(
```

Comment:
Hmmm... I was thinking schema names should be immutable, even by governance (Although technically governance can always mutate state directly).

So I was expecting `try_insert`. What are your thoughts on mutability?

Reply:
Inside `try_mutate` we are always adding the new schema id to the list via `schema_version_id.add::<T>(schema_id)?;` . Since we have an array of schema ids attached to the same name (with different versions) we have to use `mutate` , another not nice alternative is to use `get` and `set`.